eknrenderer/eknrenderer/eknr-renderer.c

```diff
@@ -0,0 +1,1408 @@
+                                  0,
+                                  escape_value_eval_callback,
+                                  NULL,
+                                  error);
+
+  if (escaped == NULL)
+    return NULL;
+
+  return g_steal_pointer (&escaped);
+}
+
+/**
+ * maybe_escape_value:
+ * @value: (transfer full): The string to escape
+ * @is_escaped: Whether or not to apply escaping
+ * @api: A #mustache_api_t, used for error reporting
+ *
+ * Based on what mustache.js does to escape HTML.
+ *
+ * Uses a regex to pick out offending characters, then replaces with
+ * HTML entities.
+ *
+ * Returns: %NULL on failure, an escaped value on success.
+ */
+static gchar *
+maybe_escape_value (gchar          *value,
+                    gboolean        is_escaped,
+                    mustache_api_t *api,
+                    void           *api_closure)
+{
+  g_autoptr(GError) error = NULL;
+  g_autofree gchar *unescaped = NULL;
+  g_autofree gchar *escaped = NULL;
+
+  if (!is_escaped)
+    return value;
+
+  /* Take an autofree reference to unescaped, it will be freed
+   * when the function returns. */
+  unescaped = value;
+  escaped = eknr_escape_html (unescaped, &error);
+
+  if (escaped == NULL)
+    {
+      (*api->error) (api, api_closure, __LINE__, error->message);
+      return NULL;
+    }
+
+  return g_steal_pointer (&escaped);
+}
+
+static uintmax_t
+_renderer_var_from_ht (mustache_api_t            *api,
+                       void                      *userdata,
+                       mustache_token_variable_t *token)
+{
+  RendererMustacheData *data = userdata;
+  g_autofree gchar *value = NULL;
+
+  /* First, if we're in a section in the value is ".", then we
+   * need to replace it with the section name. */
+  if (data->section_variable != NULL && token->text[0] == '.')
+    value = g_strdup (data->section_variable);
+  else
+    value = _lookup_in_gvariant_dict (data, token->text, api);
+
+  if (value == NULL)
+    return 0;
+
+  value = maybe_escape_value (g_steal_pointer (&value),
+                              token->escaped == 1,
+                              api,
+                              userdata);
+
+  /* Error would have been set in API. */
+  if (value == NULL)
+    return 0;
+
+  (*api->write) (api, userdata, value, strlen(value));
+  return 1;
+}
+
+static uintmax_t
+_renderer_strv_sect_from_ht (mustache_api_t           *api,
+                             void                     *userdata,
+                             mustache_token_section_t *token,
+                             GVariant                 *variant)
+{
+  RendererMustacheData *data = userdata;
+  g_autofree const gchar **strv = g_variant_get_strv (variant, NULL);
+  const gchar **iter = strv;
+
+  /* Keep track of the current section variable, then pop it once we're
+   * done */
+  const gchar *last_section_variable = data->section_variable;
+
+  /* Need to render each sub-template from here */
+  while (*iter != NULL)
```

Comment:
Maybe we want to promote the `for (iter = strv; *iter != NULL; ++iter)` style here...